akka-stream/src/main/scala/akka/stream/impl/Sources.scala

```diff
@@ -185,3 +191,143 @@ private[akka] final class SourceQueueAdapter[T](delegate: SourceQueueWithComplet
   def complete(): Unit = delegate.complete()
   def fail(ex: Throwable): Unit = delegate.fail(ex)
 }
+
+/**
+ * INTERNAL API
+ */
+private[stream] final class UnfoldResourceSource[T, S](create: () ⇒ S,
+                                                       readData: (S) ⇒ Option[T],
+                                                       close: (S) ⇒ Unit) extends GraphStage[SourceShape[T]] {
+  val out = Outlet[T]("UnfoldResourceSource.out")
+  override val shape = SourceShape(out)
+  override def initialAttributes: Attributes = DefaultAttributes.unfoldResourceSource
+
+  def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) with OutHandler {
+    lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+    var blockingStream: S = _
+    setHandler(out, this)
+
+    override def preStart(): Unit = blockingStream = create()
+
+    @tailrec
+    final override def onPull(): Unit = {
+      var resumingMode = false
+      try {
+        readData(blockingStream) match {
+          case Some(data) ⇒ push(out, data)
+          case None       ⇒ closeStage()
+        }
+      } catch {
+        case NonFatal(ex) ⇒ decider(ex) match {
+          case Supervision.Stop ⇒
+            close(blockingStream)
+            failStage(ex)
+          case Supervision.Restart ⇒
+            restartState()
+            resumingMode = true
+          case Supervision.Resume ⇒
+            resumingMode = true
+        }
+      }
+      if (resumingMode) onPull()
+    }
+
+    override def onDownstreamFinish(): Unit = closeStage()
+
+    private def restartState(): Unit = {
+      close(blockingStream)
+      blockingStream = create()
+    }
+
+    private def closeStage(): Unit =
+      try {
+        close(blockingStream)
+        completeStage()
+      } catch {
+        case NonFatal(ex) ⇒ failStage(ex)
+      }
+
+  }
+  override def toString = "UnfoldResourceSource"
+}
+
+private[stream] final class UnfoldResourceSourceAsync[T, S](create: () ⇒ Future[S],
+                                                            readData: (S) ⇒ Future[Option[T]],
+                                                            close: (S) ⇒ Future[Done]) extends GraphStage[SourceShape[T]] {
+  val out = Outlet[T]("UnfoldResourceSourceAsync.out")
+  override val shape = SourceShape(out)
+  override def initialAttributes: Attributes = DefaultAttributes.unfoldResourceSourceAsync
+
+  def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) with OutHandler {
+    lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+    var resource = Promise[S]()
```

Comment:
Opened resource is shared across read and close calls. Getting resource is asynchronous event from asyncCallback (as open function returns future). At least we cannot keep effect inside one asyncCallback.